keps/sig-network/2433-topology-aware-hints/README.md

```diff
@@ -0,0 +1,578 @@
+| Feature Gate On, Annotation == 'On', hints matching zone | Endpoints not filtered |
+| Feature Gate Off, Annotation == 'Auto', hints matching zone | Endpoints not filtered |
+| Feature Gate On, Annotation == 'Auto', no hints matching zone | Endpoints not filtered |
+
+### Observability
+We can reuse some of the metrics of EndpointSlice Controller that we already
+have in the current version to observe the changes of endpoints (addition,
+deletion and update). Meanwhile we can add more metrics to have a glimpse of
+different approaches.
+
+- `endpoint_slice_controller/endpointslices_changed_per_sync`
+- `endpoint_slice_controller/syncs`
+
+```
+const SubSystem = "endpoint_slice_controller"
+
+// This metric observes churn of EndpointSlices per sync
+EPSChangedPerSync = metrics.NewHistogramVec(
+  &metrics.HistogramOpts{
+	  Subsystem: Subsystem,
+    Name: "endpointslices_changed_per_sync",
+    Help: "Number of EndpointSlices be changed on each Service sync",
+  },
+  []string{"approach"}, // either "random" or "auto"
+)
+
+// EndpointSliceSyncs tracks the number of sync operations the controller runs along with their result.
+EndpointSliceSyncs = metrics.NewCounterVec(
+  &metrics.CounterOpts{
+    Subsystem:      EndpointSliceSubsystem,
+    Name:           "syncs",
+    Help:           "Number of EndpointSlice syncs",
+    StabilityLevel: metrics.ALPHA,
+  },
+  []string{"result"}, // either "success" or "failure"
+)
+
+```
+
+### Graduation Criteria
+- Alpha should provide basic functionality covered with tests described above.
+
+### Version Skew Strategy
+This KEP requires updates to both the EndpointSlice Controller and kube-proxy.
+Thus there could be two potential version skew scenarios:
+1. EndpointSlice Controller falls back to current behavior that does not
+   support labeling EndpointSlices. In this case, kube-proxy will still work
+   because EndpointSlices will not include topology hints.
+2. Kube-Proxy falls back to current behavior that does not support topology
+   hints in EndpointSlices. In this case, kube-proxy will continue to consume
+   all endpoints. This will not be an issue, it simply won't be taking advantage
+   of the new controller functionality.
+
+Each scenario described above will end up behaving as if this feature is not
+enabled even if the Service annotation has been set.
+
+## Production Readiness Review Questionnaire
+
+### Feature Enablement and Rollback
+
+* **How can this feature be enabled / disabled in a live cluster?**
+  - [x] Feature gate (also fill in values in `kep.yaml`)
+    - Feature gate name: TopologyAwareHints
+    - Components depending on the feature gate:
+      - kube-controller-manager
+      - kube-proxy
+
+* **Does enabling the feature change any default behavior?**
+  No.
+
+* **Can the feature be disabled once it has been enabled (i.e. can we roll back
+  the enablement)?**
+  Yes. It can easily be disabled universally by turning off the feature gate or
```

Comment:
when it will be GA it will be always enabled?

Reply:
Good question. Although the feature gate would be enabled by default, it would still require per-Service opt-in. I think we could potentially change the default value of this annotation with a follow up KEP. This is hidden a bit further up in the KEP:

>  A future KEP will explore changing the default behavior from `Disabled` to `Auto` with a new feature gate.

Reply:
:+1: